### DISCUSSION TITLE
MiniMisc.setup_auto_root not changing directory on buffer open

### DISCUSSION THREAD

**Discussion by Normanras** [nvim-mini/mini.nvim#1337] [Q&A]
I can't quite figure this one out. I've read through the docs and see that this runs `find_root` and then runs a command to `chdir` of the current working directory of the buffer. Below is what I expect to happen, can someone please correct me if this is wrong?
* Terminal on path `/user/me/`
* Open `nvim`
* `:pwd` yields `/user/me`
* Open new buffer located at `/user/me/path/file.md`
* `:pwd` should yield `/user/me/path`

However, on that last bullet point, `:pwd` is still returning `/user/me`. Is that to be expected?

Here's my setup, using lazy. All my other mini plugins are setup the same way and seem to work as it is written in mini's docs.

```
{
	'echasnovski/mini.misc', version = '*',
		config = function()
		require('mini.misc').setup({
			make_global = {'setup_auto_root','find_root'},
		})
	end
},
```

Thanks!

**Comment by echasnovski** [comment]
> * `:pwd` should yield `/user/me/path`
> 
> However, on that last bullet point, `:pwd` is still returning `/user/me`. Is that to be expected?

Without any further information, I'd say it is expected. Root directory is [by definition](https://github.com/echasnovski/mini.nvim/blob/7b4d5d48b6b5a75009d63f8f3e4ef4819b7e8139/doc/mini-misc.txt#L151-L153) is an ancestor directory that contains at least one of pre-defined file system entries (i.e. file or directory), which [by default are '.git' and 'Makefile'](https://github.com/echasnovski/mini.nvim/blob/7b4d5d48b6b5a75009d63f8f3e4ef4819b7e8139/doc/mini-misc.txt#L165). So unless '/user/me/path' or '/user' contain one of the entries, this is expected, as it seems to have not found a root directory so falls back to preserving the current working directory.

**Reply by Normanras** [reply]
Got it, so those links help and appreciate you explaining that. What is the use case you're envisioning for the `setup_auto_root` then? Perhaps that's what I'm fundamentally missing.

**Reply by echasnovski** [reply]
> What is the use case you're envisioning for the `setup_auto_root` then?

Automatic setting current working directory to the directory of the file's "project". Which is usually a Git repo, but can be adjusted to contain any file that indicates that its parent directory is a "project root" (like '.root', '.svn', 'package.json', etc.). This way I can open a file from another project and use all features that rely on current working directory (like `:Pick files`, `:lua MiniFiles.open()`, etc.).

**Reply by Normanras** [reply]
Ah, understood. And this option is a per-buffer cwd, correct?

So my use case is slightly different then. I have a directory which contains each of my clients directories and their files, which are not "projects" but just directories of various files needed for servicing those clients - so something like `/dir/client1`, `/dir/client2`, etc. I want my terminal to remain at `/dir` but that buffer to be at `../client1`, and so forth. Especially since I may have multiple clients files open at the same time, using `Pick` in each buffer within that client's files would be super helpful. I'm sure theres a custom autocmd I can write to get this to work. Appreciate your quick replies!

**Reply by echasnovski** [reply]
> Ah, understood. And this option is a per-buffer cwd, correct?

That's a bit of a tricky question. The current working directory can (more or less) be viewed as a global value of the current Neovim instance. It can be local to tabpage and window, but those are rarely used (see  `:h chdir()` and `:h current-directory` for more details). One thing to note here is that this is a concept local to currently running Neovim instance. If you change its current working directory and then exit, the terminal's directory from which `nvim` was executed would not change.

*But* `MiniMisc.setup_auto_root()` sets up autocommands that apply `vim.fn.chdir()` whenever a current buffer is changed, effectively making current working directory have a buffer-local value.

> So my use case is slightly different then.

The `:h 'autochdir'` option does what you describe here: "When on, Vim will change the current working directory ... to the directory containing the file which was opened or selected.". But not sure what you'd expect if the file deeply inside '/dir/' is opened. For example, what would you expect the current working directory be for file '/dir/client1/sub/dir/file'?

However, I'd personally suggest sticking with `MiniMisc.setup_auto_root()` and either use [callable `names` argument](https://github.com/echasnovski/mini.nvim/blob/7b4d5d48b6b5a75009d63f8f3e4ef4819b7e8139/doc/mini-misc.txt#L135) or [callable `fallback` argument](https://github.com/echasnovski/mini.nvim/blob/7b4d5d48b6b5a75009d63f8f3e4ef4819b7e8139/doc/mini-misc.txt#L136). The difference is mostly subtle, but might be useful:
- The function `names` (if written correctly) might first check if the current file has a special ancestor directory (the '/dir/' directory in your example) and set cwd to its child. Or it can check if the directory is exactly a child of target '/dir/' directory.
    If file is not inside that special directory, use usual "project finding".
- The function `fallback` might be simpler to implement and will have the similar "fallback" behavior (check if inside target special directory and return its 'client' subdirectory), but it will applied *after* a regular project finding is finished. This might be an issue if target special directory is inside a Git repo or another "project-like" directory.

Even another alternative (which I'd personally favor to consider) is to actually identify which use cases need this "special target current working directory" approach and tweak *them* instead. So for `:Pick` it can be executing custom picker with `source.cwd` set to directory you want to search with. Mapped to a separate keys it will make an independent action which would not interfere with a usual "find in current working directory" cases.
